include/pybind11/detail/internals.h

```diff
@@ -39,7 +39,11 @@
 #    if PY_VERSION_HEX >= 0x030C0000 || defined(_MSC_VER)
 // Version bump for Python 3.12+, before first 3.12 beta release.
 // Version bump for MSVC piggy-backed on PR #4779. See comments there.
-#        define PYBIND11_INTERNALS_VERSION 5
+#        ifdef Py_GIL_DISABLED
+#            define PYBIND11_INTERNALS_VERSION 6
```

Comment:
Unless #5296 is merged first, this will lead to incompatibility between extension modules compiled with or without `Py_GIL_DISABLED` defined.

I just only now realize ... Huston we have a problem?

We already added `pymutex mutex` to `struct internals` but didn't change `PYBIND11_INTERNALS_VERSION`. Ouch? Sorry I missed that before.

What do we want here? I'm very uncertain, @colesbury could you please advise?

```C++
struct internals {
#if PY_VERSION_HEX >= 0x030D0000
    pymutex mutex;
    pymutex exception_translator_mutex;
#endif
...
```

Could that work? The idea here is that `struct internals` is the same with and without `Py_GIL_DISABLED` defined.

Reply:
Free-threading is experimental so changing the ABI before was fine. Now I think packages might be beginning to ship, so probably not such a great idea to change without a bump. But not fond of these confusing bumps. The ABI is currently a mess, by the way - the Windows only bump broke conda-forge and the only solution was to pin <2.12, which now is breaking packages (see the feedstock for more info).

Will the internals version matter after #5296? Could we just basically do away with the concept, and only do compiler/stdlib, etc?

Reply:
> Could we just basically do away with the concept

Not entirely, **I think**. The internals versions still need to match for the inheritance functionality to work correctly. (I'd probably need a day or two to figure out conclusively if that's fixable or just fundamentally impossible.)

But the more I think about it, the more I'm getting worried about having a different `struct internals` with and without `Py_GIL_DISABLED` defined:

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/internals.h#L177-L180

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/internals.h#L185-L190

What happens if there is one extension built with the GIL enabled, another built with free-threading, and both are imported into the same process?

Currently we only have this:

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/common.h#L390-L405

Do we need to also check for `Py_GIL_DISABLED` compatibility there?

Or do we have to do something here?

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/internals.h#L324-L332

@colesbury

Reply:
> What happens if there is one extension built with the GIL enabled, another built with free-threading, and both are imported into the same process?

You can't have extensions compiled with and without `Py_GIL_DISABLED` in the same process. An extension might require the GIL to be dynamically enabled, but that's different: that's a matter of calling (or not calling) `PyUnstable_Module_SetGIL()`. The extension will still need to be compiled with `Py_GIL_DISABLED` to be loaded in 3.13t.

In other words:

* `Py_GIL_DISABLED` is part of the extension ABI, similar to `Py_VERSION_HEX`
* It's part of the wheel tag (i.e., the "t" in `cp313-cp313t`), just like the version number

Reply:
Or as @henryiii wrote in https://github.com/pybind/pybind11/pull/5148#discussion_r1619348108: "A free-threaded build can't talk to a non-free-threaded build".

Reply:
@vfdev-5 that's not related to the `struct internals`, `PYBIND11_INTERNALS_VERSION`, or even free-threading.

Reply:
I agree, I should have built using same python version, e.g. 3.13t and 3.13.
Doing so, above example is giving either a segfault:
```bash
$ gdb --args python3.13t main.py

Starting program: /usr/bin/python3.13t main.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7be1f0b in pybind11::dict::dict(pybind11::object&&) () from /pybind11_playground/ft_or_not_ft/test_gil.so
```
or 
```bash
$ python3.13 main.py
Traceback (most recent call last):
  File "/pybind11_playground/ft_or_not_ft/main.py", line 4, in <module>
    import test_ft
ImportError: /pybind11_playground/ft_or_not_ft/test_ft.so: undefined symbol: _Py_DecRefShared
```

If this is again unrelated sorry for the noise.

Reply:
Thanks a lot! That really helps me understand the situation better. So I think what we need to do is work on this code:

https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/common.h#L390-L405

We want to catch if there is a mismatch between the Python interpreter kind (gil, free-threading) and the way the extension was built.

**In a separate PR.**

Reply:
> Now I think packages might be beginning to ship, so probably not such a great idea to change without a bump. But not fond of these confusing bumps.

Yeah ... "version bump chaos"

Trying to summarize how I understand the situation:

* The internals version wasn't changed for Python 3.13t — but that's ok because there were no released packages anyway that could have been built with a non-matching `struct internals`.

* Assuming there are released packages for 3.13t now, yes, it would indeed be safest to bump the internals version for 3.13t. — But pegging that on `Py_GIL_DISABLED` will make it very difficult for the general user community to reason about the internals version.

If this PR gets merged as is, we'll have internals version

* `4` for Python < 3.12 if `_MSC_VER` is not defined.
* `5` for Python 3.12, or 3.13, or if `_MSC_VER` is defined.
* `6` for Python 3.13t only.

Who can (or wants to) understand that? :-)

And then we also still have #5339 (smart_holder) and #5280 (native enum) that I'd really like to see merged.

I believe this is a situation where the maintainers need to step up to balance "making progress" with "minimizing confusion".

@henryiii wrote:

> Free-threading is experimental

That makes me think, let's not mess with the internals version in this PR:

* Maybe a small number of users **experimenting** will stumble,
* but a much larger number of **production** users will not have to understand the weirdness summarized above.

WDYT?

Looking beyond that question, I feel it'd be best to approach this strategically:

* Merge #5296 asap (only needs a small amount of work).
* Merge #5339, #5280 as closely together in time as possible and start the pybind11 v3 series, with internals version **`6`** unconditionally.

That's a discussion far beyond this PR, but I'm mentioning it here because it would be another reason to not mess with the internals version here.

Reply:
> That makes me think, let's not mess with the internals version in this PR

I'm concerned that landing this PR without changing the version number will cause problems not just for users but also package maintainers (e.g., scipy, PyTorch) that use pybind11 and support free-threading. The failure modes for different internals can be complicated -- in this case, I think you run the risk of treating uninitialized memory like a mutex.

I think the approach in this PR is a good long term fix, but if the version bump is the issue we can first pursue a different approach that avoids changing the internals struct (and consider landing this change later when #5280 is ready).

Reply:
First priority, I want to be helpful. Second minimize potential for confusion.

Unfortunately I'm technically somewhat handicapped at the moment (no Linux workstation), but let me try to get #5296 merged asap, with @henryiii's formal approval, to be sure we're on the same.

After that version bumps will have a lot less impact, and it won't matter as much that people fully understand the numbering.